src/datatablerow/utils.jl

```diff
@@ -0,0 +1,190 @@
+# Rows grouping.
+# Maps row contents to the indices of all the equal rows.
+# Used by groupby(), join(), nonunique()
+immutable RowGroupDict{T<:AbstractDataTable}
+    dt::T                 # source data frame
+
+    ngroups::Int          # number of groups
+
+    rhashes::Vector{UInt} # row hashes
+    gslots::Vector{Int}   # hashindex -> index of group-representative row
+
+    groups::Vector{Int}   # group index for each row
+    rperm::Vector{Int}    # permutation of row indices that sorts them by groups
+
+    starts::Vector{Int}   # starts of ranges in rperm for each group
+    stops::Vector{Int}    # stops of ranges in rperm for each group
+end
+
+# "kernel" functions for hashrows()
+# adjust row hashes by the hashes of column elements
+function hashrows_col!(h::Vector{UInt}, v::AbstractVector)
+    @inbounds for i in eachindex(h)
+        h[i] = hash(v[i], h[i])
+    end
+    h
+end
+
+function hashrows_col!{T}(h::Vector{UInt}, v::AbstractVector{T})
+    @inbounds for i in eachindex(h)
+        h[i] = isnull(v[i]) ?
+               hash(NULL_MAGIC, h[i]) :
+               hash(v[i], h[i])
+    end
+    h
+end
+
+# should give the same hash as AbstractVector{T}
+function hashrows_col!{T}(h::Vector{UInt}, v::AbstractCategoricalVector{T})
+    # TODO is it possible to optimize by hashing the pool values once?
```

Comment:
I've found a recent discussion in PostgreSQL about this exact issue: combining hashes of individual elements:
https://www.postgresql.org/message-id/CAEepm%3D3rdgjfxW4cKvJ0OEmya2-34B0qHNG1xV0vK7TGPJGMUQ%40mail.gmail.com

They were (and are still) using a quite simple algorithm, with a bit shift followed by a `xor`. So that's would be a reasonable start for us. In the discussion, potential improvements are mentioned: [boost::hash_combine](http://www.boost.org/doc/libs/1_55_0/doc/html/hash/reference.html#boost.hash_combine), which is widely used and appears to be designed for the problem at hand (see [this discussion](http://stackoverflow.com/questions/4948780/magic-number-in-boosthash-combine for more details). I suggest we use that solution.

Also useful: http://eternallyconfuzzled.com/tuts/algorithms/jsw_tut_hashing.aspx